### PR TITLE
[Flink-12208][table-planner-blink] Support translation from StreamExecSort / TemporalSort / SortLimit/ Limit to StreamTransformation.

### PR DESCRIPTION
**Pull request by beyond1920**
## What is the purpose of the change

Support translation from StreamExecSort / TemporalSort / SortLimit/ Limit to StreamTransformation.
1. StreamExecSort handles query which requires sort on non-time fields, which caches all inputs elements, sorts all cached elements when the bounded-stream is finished, outputs sorted elements one by one then.
*Notes:* it's only used for testing with bounded source now. If a query is converted to this node in product environment, an exception will be thrown.
2. StreamExecTemporalSort handles query which requires sort on EventTime/ ProcTime fields
3. StreamExecLimit  could handles query which only takes `limit` elements from offset. 
*Notes:* StreamExecLimit could only support take limited elements currently, that is, `fetch` must be present in query.
4. StreamExecSortLimit  handles query which not only requires sort, but also only takes `limit` elements from offset. 
*Notes:* StreamExecSortLimit could only support take limited elements currently, that is, `fetch` must be present in query.

## Brief change log

* Introduce StreamSortOperator to process sort on Non-time fields.
* Introduce OnlyRowTimeSortOperator/RowTimeSortOperator/ProcTimeSortOperator to process sort on time fields.
*  StreamExecSort / StreamExecTemporalSort / StreamExecSortLimit/ StreamExecLimit implements StreamExecNode.

## Verifying this change

UT && IT

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala [hunk 1]
```diff
@@ -72,4 +89,78 @@ class StreamExecLimit(
...
+    val generateRetraction = StreamExecRetractionRules.isAccRetract(this)
+    val tableConfig = tableEnv.getConfig
+    val minIdleStateRetentionTime = tableConfig.getMinIdleStateRetentionTime
+    val maxIdleStateRetentionTime = tableConfig.getMaxIdleStateRetentionTime
+
+    // rankStart begin with 1
+    val rankRange = new ConstantRankRange(limitStart + 1, limitEnd)
+    val rankType = RankType.ROW_NUMBER
+    val outputRankNumber = false
+    // TODO Use RankFunction underlying StreamExecLimit currently
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala]: Why we use RankFunction instead of a codegened sorter just like StreamExecSort did? Same question to StreamExecSortLimit

Comment (beyond1920) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala]: Because SortLimit and Limit takes `limited` elements from `offset`, which is very similar to TopN, and could use AppendRankFunction or UpdateRankFunction under some as optimization in some scenarios. BTW, maybe RankFunction is not a good name, it's only for TopN.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/BaseTemporalSortOperator.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		InternalTimerService<VoidNamespace> internalTimerService = getInternalTimerService("user-timers",
+				VoidNamespaceSerializer.INSTANCE,
+				this);
+		timerService = new SimpleTimerService(internalTimerService);
+		collector = new TimestampedCollector<>(output);
+	}
+
+	@Override
+	public void onEventTime(InternalTimer<BaseRow, VoidNamespace> timer) throws Exception {
+		throw new UnsupportedOperationException("Now Sort only is supported based processing time here!");
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/BaseTemporalSortOperator.java]: I think we can un-implement the `onEventTime` and `onProcessintTime` in the base class and let the subclasses to explicitly throw exception.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/LimitOperator.java [hunk 3]
```diff
@@ -26,7 +26,7 @@
 import org.apache.flink.util.Collector;
 
 /**
- * Operator for limit.
+ * Operator for batch limit.
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/sort/LimitOperator.java]: Can we relocate the operators into `stream` or `batch` sub-package`? 

Currently, the classes names are quite confusing. IMO, adding a comment doesn't help a lot here.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/deduplicate/DeduplicateFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.flink.table.runtime.util.BinaryRowKeySelector;
+import org.apache.flink.table.runtime.util.GenericRowRecordSortComparator;
+import org.apache.flink.table.runtime.util.StreamRecordHelper;
+import org.apache.flink.table.type.InternalTypes;
+import org.apache.flink.table.typeutils.BaseRowTypeInfo;
+
+/**
+ * Base class of tests for all kinds of DeduplicateFunction.
+ */
+abstract class DeduplicateFunctionTest {
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/deduplicate/DeduplicateFunctionTest.java]: Rename to `DeduplicateFunctionTestBase`? Because this is not a test class.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/rank/BaseTopNFunctionTest.java [hunk 5]
```diff
@@ -41,14 +44,10 @@
...
-import static org.apache.flink.table.runtime.util.StreamRecordUtils.deleteRecord;
-import static org.apache.flink.table.runtime.util.StreamRecordUtils.record;
-import static org.apache.flink.table.runtime.util.StreamRecordUtils.retractRecord;
-
 /**
- * Base Tests for all subclass of {@link AbstractRankFunction}.
+ * Base Tests for all subclass of {@link AbstractTopNFunction}.
  */
-abstract class BaseRankFunctionTest {
+abstract class BaseTopNFunctionTest {
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/rank/BaseTopNFunctionTest.java]: I would suggest to use `TopNFunctionTestBase` as class name. Because a `*Test` suffix will be recognized as JUnit tests.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/window/WindowOperatorContractTest.java [hunk 6]
```diff
@@ -66,6 +68,8 @@
 	@Rule
 	public ExpectedException thrown = ExpectedException.none();
 
+	private StreamRecordHelper recordWrapper = new StreamRecordHelper(new RowTypeInfo(Types.STRING, Types.INT, Types.LONG));
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/window/WindowOperatorContractTest.java]: recordHelper?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala [hunk 7]
```diff
@@ -72,4 +90,78 @@ class StreamExecLimit(
...
+    val inputTransform = getInputNodes.get(0).translateToPlan(tableEnv)
+      .asInstanceOf[StreamTransformation[BaseRow]]
+
+    val outputRowTypeInfo = FlinkTypeFactory.toInternalRowType(getRowType).toTypeInfo
+    val ret = new OneInputTransformation(
+      inputTransform,
+      s"Limit(offset: $limitStart, fetch: ${fetchToString(fetch)})",
+      operator,
+      outputRowTypeInfo,
+      inputTransform.getParallelism)
```
[the remark this answers is not in the document]
Comment (beyond1920) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala]: Since we limit the exchange must be singleton before StreamExecLimit, SortLimit in the rules, is it still necessary to force parallelism to 1 here.

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/stream/StreamExecLimit.scala]: Yes, I think so. The parallelism of singleton `PartitionTransformation` may not be 1.

**Comment by beyond1920** [comment]
@wuchong  thanks for suggestions. I have updated the pr. Please take a look at it.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/RankITCase.scala [hunk 8]
```diff
@@ -1095,34 +1070,35 @@ class RankITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mode
     assertEquals(updatedExpected.sorted, sink.getUpsertResults.sorted)
   }
 
-  // FIXME
-  @Ignore("Enable after agg implements ExecNode")
   @Test
   def testMultipleRetractTopNAfterAgg(): Unit = {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/RankITCase.scala]: I would like to ignore this test and revert the changes, because current implementation lost the original test purpose, i.e. "test multiple retract sink after topN".

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/LimitITCase.scala [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      ("fruit", 5, 22))
+
+    val t = env.fromCollection(data).toTable(tEnv, 'category, 'shopId, 'num)
+    tEnv.registerTable("T", t)
+
+    val sql = "SELECT * FROM T OFFSET 2"
+
+    val sink = new TestingRetractSink
+    tEnv.sqlQuery(sql).toRetractStream[Row].addSink(sink).setParallelism(1)
+    env.execute()
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/LimitITCase.scala]: This can be simplified to 

```scala
    thrown.expect(...)
    thrown.expectMessage(...)
    tEnv.sqlQuery(sql).toRetractStream[Row]
```

`toRetractStream` will call translate plan, we don't need to call `execute()` here I think.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/SortITCase.scala [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+import org.junit.{Ignore, _}
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+import scala.collection.mutable
+
+@RunWith(classOf[Parameterized])
+class SortITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mode) {
+
+  @Test(expected = classOf[TableException])
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/runtime/stream/sql/SortITCase.scala]: The same as above, please use `thrown` to check and we don't need to call `execute()`.

**Comment by wuchong** [comment]
@flinkbot  approve-until quality

**Comment by wuchong** [comment]
LGTM +1

**Comment by wuchong** [comment]
@flinkbot approve all

**Comment by wuchong** [comment]
Merging...
